File: tests/test_angle_precision.py
"""bd a0w (M5): the ANGLE-PRECISION lever (Ross-Selinger / gridsynth) — Python.

End-to-end gates for the lever; the pure-quantizer and clamp unit tests live in
``tests/test_angle_precision.c``.

The lever models finite synthesis accuracy of the QTG's per-variable rotation
``R_y(theta_i)``: gridsynth reaches an absolute angle accuracy ``eps`` radians
at ``~3*log2(1/eps)`` T-gates, so a coarser angle buys a cheaper circuit. The
input is always ``eps`` (radians); bits and T-counts are reporting quantities.

What is pinned here:

  * param validation + per-phase resolution (``sat_/opt_sat_/opt_`` come free
    from ``PHASE_PARAM_SUFFIXES``);
  * NEGATIVE CONTROL — the lever unset (and ``dither`` armed with no ``eps``)
    is BIT-FOR-BIT the pre-a0w solve: identical objective, solution array AND
    oracle-indexed history;
  * a very fine ``eps`` reproduces the exact-angle arm (the sweep's negative
    control at solver level);
  * DETERMINISM (NORTHSTAR §8) under a fixed seed + single worker with the
    lever ON, dither both OFF and ON;
  * BOUNDED DECISIONS (§1.7) survive a grid coarse enough to round theta to
    zero at the n=3000 operating point — the solve stays feasible-capable and
    never crashes;
  * the realized opt-phase radius from ``result.branch_diagnostics`` tracks the
    PREDICTED quantized radius ``r_q = n*sin^2(theta_q/2)`` — i.e. the lever
    really moves the neighborhood the way the angle model says it does.
"""
import math

import pytest

pytest.importorskip("cbqs")

from cbqs.Model import Model
from cbqs.Constants import MAXIMIZE
from cbqs.phase_params import (
    PhaseParamResolver, DEFAULTS, PHASE_PARAM_SUFFIXES,
    angle_precision_bits, t_count_per_rotation, t_count_per_qtg_application,
)


def _build(n=40, cap=20):
    """Small feasible knapsack-style model (all-zeros feasible -> enters opt)."""
    m = Model()
    x = m.add_variables(n)
    m.set_objective(sum(x[i] for i in range(n)), MAXIMIZE)
    m.add_constraint(sum(x[i] for i in range(n)) <= cap)
    m.close()
    return m


def _solve(params, *, seed=7, M=1500, workers=1, n=40, cap=20):
    m = _build(n, cap)
    m.seed = seed
    m.set_param("M", M)
    m.set_param("num_workers", workers)
    m.set_param("track_history", True)
    for k, v in params.items():
        m.set_param(k, v)
    r = m.solve()
    hist = tuple((float(a), int(b)) for a, b in (r.history or []))
    return r, hist


def _fingerprint(r, hist):
    return (r.objective, tuple(int(b) for b in r.solution), bool(r.feasible), hist)


# --------------------------------------------------------------------------- #
# param validation + resolution
# --------------------------------------------------------------------------- #

class TestValidation:
    @pytest.mark.parametrize("key", ["angle_precision_eps",
                                     "sat_angle_precision_eps",
                                     "opt_sat_angle_precision_eps",
                                     "opt_angle_precision_eps"])
    @pytest.mark.parametrize("bad", [0.0, -1.0, -1e-9])
    def test_rejects_nonpositive_eps(self, key, bad):
        """<= 0 is the OFF switch, expressed as None — never as a set value."""
        m = _build()
        with pytest.raises(ValueError):
            m.set_param(key, bad)

    def test_accepts_positive_eps_and_dither(self):
        m = _build()
        m.set_param("angle_precision_eps", 0.01)
        m.set_param("angle_precision_dither", True)
        m.set_param("opt_angle_precision_eps", 0.002)
        assert m._params["angle_precision_eps"] == 0.01
        assert m._params["angle_precision_dither"] is True
        assert m._params["opt_angle_precision_eps"] == 0.002

    @pytest.mark.parametrize("key", ["angle_precision_dither",
                                     "sat_angle_precision_dither",
                                     "opt_angle_precision_dither"])
    @pytest.mark.parametrize("bad", ["false", "no", "off", "", [1], 2, -1, 0.5])
    def test_dither_rejects_non_boolean(self, key, bad):
        """§2.1 fail loud: plain ``bool`` coercion makes every non-empty string
        truthy, so ``set_param(..., 'false')`` would silently ARM the flag and
        switch which PHYSICAL synthesis model an arm runs — corrupting a whole
        sweep with no error. Only real booleans and 0/1 are accepted."""
        m = _build()
        with pytest.raises(ValueError):
            m.set_param(key, bad)

    @pytest.mark.parametrize("good,expected", [(True, True), (False, False),
                                               (1, True), (0, False)])
    def test_dither_accepts_bools_and_0_1(self, good, expected):
        m = _build()
        m.set_param("angle_precision_dither", good)
        assert m._params["angle_precision_dither"] is expected

    def test_unknown_param_rejected(self):
        m = _build()
        with pytest.raises(ValueError):
            m.set_param("angle_precision_bits", 4)

    def test_default_is_off(self):
        m = _build()
        assert m._get_effective("angle_precision_eps") is None
        assert m._get_effective("angle_precision_dither") is False


class TestPhaseResolution:
    def test_suffixes_registered(self):
        assert "angle_precision_eps" in PHASE_PARAM_SUFFIXES
        assert "angle_precision_dither" in PHASE_PARAM_SUFFIXES
        assert DEFAULTS["angle_precision_eps"] is None
        assert DEFAULTS["angle_precision_dither"] is False

    def test_phase_specific_beats_unprefixed(self):
        store = {"angle_precision_eps": 0.05, "opt_angle_precision_eps": 0.001,
                 "angle_precision_dither": True}
        res = PhaseParamResolver(store, DEFAULTS)
        assert res.resolve("sat", "angle_precision_eps") == 0.05
        assert res.resolve("opt_sat", "angle_precision_eps") == 0.05
        assert res.resolve("opt", "angle_precision_eps") == 0.001
        assert res.resolve("opt", "angle_precision_dither") is True

    def test_unset_resolves_to_off(self):
        res = PhaseParamResolver({}, DEFAULTS)
        for phase in ("sat", "opt_sat", "opt"):
            assert res.resolve(phase, "angle_precision_eps") is None
            assert res.resolve(phase, "angle_precision_dither") is False


class TestReportingHelpers:
    """Bits / T-counts are DERIVED for reading results, never solver inputs."""

    def test_bits_matches_uniform_grid(self):
        # a b-bit uniform grid on [0, pi] has max rounding error pi/2**(b+1)
        for b in (2, 4, 8, 12):
            eps = math.pi / 2 ** (b + 1)
            assert angle_precision_bits(eps) == pytest.approx(b)

    def test_t_count_is_ross_selinger(self):
        assert t_count_per_rotation(0.01) == pytest.approx(3 * math.log2(100))
        assert t_count_per_qtg_application(0.01, 3000) == pytest.approx(
            3000 * 3 * math.log2(100))

    @pytest.mark.parametrize("bad", [0.0, -1.0, None])
    def test_reporting_helpers_fail_loud(self, bad):
        with pytest.raises(ValueError):
            angle_precision_bits(bad)
        with pytest.raises(ValueError):
            t_count_per_rotation(bad)


# --------------------------------------------------------------------------- #
# NEGATIVE CONTROL: lever OFF == pre-a0w solve, bit-for-bit
# --------------------------------------------------------------------------- #

class TestOffIsBitForBit:
    def test_dither_without_eps_is_a_no_op(self):
        """`dither` alone must not perturb anything — eps is the only switch."""
        base = _fingerprint(*_solve({}))
        for dither in (False, True):
            got = _fingerprint(*_solve({"angle_precision_dither": dither}))
            assert got == base

    def test_per_phase_dither_without_eps_is_a_no_op(self):
        base = _fingerprint(*_solve({}))
        got = _fingerprint(*_solve({"sat_angle_precision_dither": True,
                                    "opt_sat_angle_precision_dither": True,
                                    "opt_angle_precision_dither": True}))
        assert got == base

    def test_fine_eps_is_numerically_near_exact_but_NOT_bit_for_bit(self):
        """The sweep's `negctl` arm is NEAR-exact, and the docs must say so.

        `eps <= 0` is the only bit-for-bit control (the quantizer block is
        skipped entirely -- pinned by the two tests above). A tiny positive eps
        still runs the round trip `1 - sin^2(asin(sqrt(p)))`, which is lossy at
        the ~1e-13 level; since sampling compares `random_num > value`, a
        perturbation that small can eventually flip a decision, so a
        "must reproduce the exact arm bit-for-bit" claim would be false and
        would fire spuriously. What IS guaranteed is the numeric bound.
        """
        for bias in (5.0, 43.0, 1498.0):          # r=4 at n=180; r=2 at n=3000
            v = (bias + 1.0) / (bias + 2.0)
            p = 1.0 - v
            theta = 2 * math.asin(math.sqrt(p))
            for eps in (1e-12, 1e-9):
                delta = 2 * eps
                theta_q = delta * round(theta / delta)
                v_q = 1.0 - math.sin(theta_q / 2) ** 2
                assert abs(theta_q - theta) <= eps
                assert abs(v_q - v) <= 4 * eps       # |dv| = |sin(theta)*dtheta|/2 <= eps
        # at least one operating point is genuinely NOT bit-identical, so the
        # weaker claim is the necessary one (guards against re-tightening it).
        v = (5.0 + 1.0) / (5.0 + 2.0)
        theta = 2 * math.asin(math.sqrt(1.0 - v))
        theta_q = 2e-12 * round(theta / 2e-12)
        assert 1.0 - math.sin(theta_q / 2) ** 2 != v


# --------------------------------------------------------------------------- #
# DETERMINISM (NORTHSTAR §8) with the lever ARMED
# --------------------------------------------------------------------------- #

class TestDeterminism:
    @pytest.mark.parametrize("dither", [False, True])
    @pytest.mark.parametrize("eps", [0.25, 0.03, 0.002])
    def test_fixed_seed_single_worker_is_reproducible(self, eps, dither):
        params = {"angle_precision_eps": eps, "angle_precision_dither": dither}
        a = _fingerprint(*_solve(params, seed=11, workers=1))
        b = _fingerprint(*_solve(params, seed=11, workers=1))
        assert a == b

    def test_dither_changes_the_trajectory(self):
        """Dither must actually DO something — otherwise the on/off arms of the
        sweep would be measuring the same thing."""
        eps = 0.25
        off = _fingerprint(*_solve({"angle_precision_eps": eps,
                                    "angle_precision_dither": False}, seed=11))
        on = _fingerprint(*_solve({"angle_precision_eps": eps,
                                   "angle_precision_dither": True}, seed=11))
        assert off != on


# --------------------------------------------------------------------------- #
# BOUNDED DECISIONS (§1.7) end-to-end at a coarse grid
# --------------------------------------------------------------------------- #

class TestBoundedDecisionsEndToEnd:
    @pytest.mark.parametrize("dither", [False, True])
    @pytest.mark.parametrize("eps", [0.5, 0.25, 0.125, 0.0625, 0.03, 0.008, 0.001])
    def test_coarse_grid_still_solves(self, eps, dither):
        """Even a grid that rounds theta to 0 (deterministic-greedy in the naive
        implementation) must leave a well-formed, verifiable solve."""
        r, _ = _solve({"angle_precision_eps": eps,
                       "angle_precision_dither": dither}, M=600)
        assert r.objective is not None
        assert len(r.solution) == 40
        assert set(int(b) for b in r.solution) <= {0, 1}

    def test_snap_to_zero_collapses_the_radius_but_stays_well_formed(self):
        """The fail-loud case, pinned on its MEASURABLE consequence.

        A grid with eps > theta rounds the angle to zero. §1.7 then guarantees
        only that the probability stays inside (0,1) -- at the clamp floor
        p_flip = 1e-9, which is deterministic greedy in every practical sense.
        So the honest assertion is not "the stage stays exploratory" but the
        observable: the realized opt-phase Hamming radius COLLAPSES relative to
        the exact-angle arm, while the solve itself stays well-formed (a valid
        binary solution, real diagnostics, no crash). That collapse is exactly
        the cliff mechanism the a0w sweep measures.
        """
        n, r_target = 200, 2.0
        theta = 2 * math.asin(math.sqrt(r_target / n))
        assert 0.5 > theta, "precondition: eps=0.5 must round theta to zero"

        def solve(eps):
            m = _build(n=n, cap=n // 2)
            m.seed = 3
            m.set_param("M", 400)
            m.set_param("num_workers", 1)
            m.set_param("branching_radius", r_target)
            m.set_param("opt_switch_oracles", 0)
            if eps is not None:
                m.set_param("angle_precision_eps", eps)
            return m.solve()

        exact, crushed = solve(None), solve(0.5)
        for res in (exact, crushed):
            assert res.objective is not None
            assert len(res.solution) == n
            assert set(int(b) for b in res.solution) <= {0, 1}
            assert res.branch_diagnostics["n"] == n
        if not (exact.branch_diagnostics["opt_candidates"]
                and crushed.branch_diagnostics["opt_candidates"]):
            pytest.skip("no opt-phase candidates under the capped budget")
        assert crushed.branch_diagnostics["radius_mean"] < \
            0.5 * exact.branch_diagnostics["radius_mean"]


# --------------------------------------------------------------------------- #
# realized radius tracks the PREDICTED quantized radius
# --------------------------------------------------------------------------- #

def _predicted_quantized_radius(n, r, eps):
    """r_q = n*sin^2(theta_q/2) for the SYSTEMATIC grid (all variables share it).

    Uses floor(x+0.5), NOT Python's round(): C99 round() is half-away-from-zero
    while Python's is banker's rounding, and this helper exists to mirror the C
    lever exactly."""
    theta = 2 * math.asin(math.sqrt(r / n))
    delta = 2 * eps
    theta_q = delta * math.floor(theta / delta + 0.5)
    return n * math.sin(theta_q / 2) ** 2


class TestRealizedRadiusMatchesPrediction:
    """§1.5 scale-invariance diagnostic, applied to the quantized lever."""

    @pytest.mark.parametrize("eps", [0.02, 0.05])
    def test_realized_radius_ratio_matches_predicted(self, eps):
        pytest.importorskip("numpy")
        from benchmarks.scale_invariance import radius_profile

        n, r = 200, 4.0
        common = dict(n=n, index=0, seed=0, num_workers=4, M=800)
        exact = radius_profile(params={"branching_radius": r}, **common)
        quant = radius_profile(params={"branching_radius": r,
                                       "angle_precision_eps": eps}, **common)
        if exact["opt_candidates"] == 0 or quant["opt_candidates"] == 0:
            pytest.skip("no opt-phase candidates under the capped budget")

        predicted_ratio = _predicted_quantized_radius(n, r, eps) / r
        realized_ratio = quant["r_mean"] / exact["r_mean"]
        # The realized radius is f(n)-scaled in both arms, so the RATIO cancels
        # the (n-dependent) compression factor and isolates the quantization.
        assert realized_ratio == pytest.approx(predicted_ratio, rel=0.15), (
            f"eps={eps}: predicted r_q/r={predicted_ratio:.4f}, "
            f"realized {quant['r_mean']:.4f}/{exact['r_mean']:.4f}"
            f"={realized_ratio:.4f}")

    def test_coarse_grid_collapses_the_radius(self):
        """The cliff mechanism, observed directly: a grid coarser than theta/2
        rounds the angle to zero, so the realized radius collapses toward the
        clamp floor rather than staying at r."""
        pytest.importorskip("numpy")
        from benchmarks.scale_invariance import radius_profile

        n, r = 200, 4.0
        common = dict(n=n, index=0, seed=0, num_workers=4, M=800)
        exact = radius_profile(params={"branching_radius": r}, **common)
        crushed = radius_profile(params={"branching_radius": r,
                                         "angle_precision_eps": 0.5}, **common)
        if exact["opt_candidates"] == 0 or crushed["opt_candidates"] == 0:
            pytest.skip("no opt-phase candidates under the capped budget")
        assert _predicted_quantized_radius(n, r, 0.5) < 1e-6
        assert crushed["r_mean"] < 0.5 * exact["r_mean"]

File: tests/test_angle_precision.c
/*
 * tests/test_angle_precision.c -- bd a0w (M5): the ANGLE-PRECISION lever.
 *
 * On hardware the QTG's per-variable rotation R_y(theta_i) is not exact: it is
 * synthesized by Ross-Selinger / gridsynth to an absolute accuracy `eps` at a
 * T-count ~3*log2(1/eps). This lever models that inside BranchingFunction by
 * quantizing THETA (never the probability `value`): the physical knob is the
 * angle, and a uniform grid in probability space is unfaithful and far too
 * coarse near p ~ 0, which is exactly where CBQS operates (p_flip = r/n).
 *
 *   p_flip = 1 - value = sin^2(theta/2)      (the QTG amplitude convention)
 *   systematic : theta_q = 2*eps * round(theta / (2*eps))   [|dtheta| <= eps]
 *   dithered   : theta_q = theta + eps * u(index)           [|dtheta| <= eps]
 *
 * These tests pin, in order of load-bearingness:
 *   (1) eps <= 0 is the OFF switch and is BIT-FOR-BIT the pre-a0w path -- the
 *       NORTHSTAR §8 branching golden 6/7 at bias=5 must survive exactly;
 *   (2) the synthesis-accuracy contract |theta_q - theta| <= eps in both modes;
 *   (3) BOUNDED DECISIONS (§1.7): every eps, including a grid so coarse it
 *       snaps theta to 0 at the n=3000 operating point, must leave the returned
 *       probability strictly inside (0,1) -- a naive grid would force greedy;
 *   (4) the dither is a DETERMINISTIC pure function of the variable index
 *       (reproducible under a fixed seed -> the §8 determinism baseline holds).
 */
#include <stdarg.h>
#include <stddef.h>
#include <setjmp.h>
#include <cmocka.h>
#include <math.h>
#include <string.h>

#include "Branching.h"
#include "solver_ctx.h"
#include "definitions.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

/* n=3000 at the shipped radius r~2: bias = n/r - 2 = 1498, p_flip = 2/3000. */
#define BIAS_N3000_R2 1498.0
#define P_N3000_R2    (2.0 / 3000.0)

static int ap_setup(void **state) {
    solver_ctx_t *ctx = solver_ctx_create();
    if (ctx == NULL) return -1;
    *state = ctx;
    return 0;
}

static int ap_teardown(void **state) {
    solver_ctx_free((solver_ctx_t *)*state);
    *state = NULL;
    return 0;
}

/* Reference: the angle the lever quantizes, from the un-quantized probability. */
static double theta_of(double value) { return 2.0 * asin(sqrt(1.0 - value)); }

/* ============================================================
 * (1) OFF is bit-for-bit the pre-a0w path
 * ============================================================ */

/* Default-constructed stats must have the lever OFF. */
static void test_defaults_are_off(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    assert_true(ctx->branching_stats.angle_precision_eps == 0.0);
    assert_int_equal(ctx->branching_stats.angle_precision_dither, 0);
    assert_true(ctx->branching_stats_sat.angle_precision_eps == 0.0);
    assert_true(ctx->branching_stats_opt_sat.angle_precision_eps == 0.0);
    assert_true(ctx->branching_stats_opt.angle_precision_eps == 0.0);
}

/* THE golden (NORTHSTAR §8): bias=5, no weights => exactly 6/7, lever off. */
static void test_golden_6_over_7_preserved(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    solver_ctx_set_bias(ctx, 5.0);
    double v = BranchingFunction(0, 0, 0, 0, &ctx->branching_stats);
    assert_true(v == 6.0 / 7.0);   /* bit-for-bit, not assert_float_equal */
}

/* eps <= 0 (and the untouched default) are the same OFF switch. */
static void test_nonpositive_eps_is_off(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    solver_ctx_set_bias(ctx, 5.0);
    double base = BranchingFunction(3, 0, 0, 0, &ctx->branching_stats);
    for (int dither = 0; dither <= 1; dither++) {
        solver_ctx_set_angle_precision(ctx, 0.0, dither);
        assert_true(BranchingFunction(3, 0, 0, 0, &ctx->branching_stats) == base);
        solver_ctx_set_angle_precision(ctx, -1.0, dither);
        assert_true(BranchingFunction(3, 0, 0, 0, &ctx->branching_stats) == base);
    }
}

/* The OFF path is byte-identical for the WHOLE (bit_S, bit_T) truth table and
 * through the per-variable weights path, not just the golden cell. */
static void test_off_matches_baseline_everywhere(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    double w[4] = {0.7, -0.3, 0.0, 1.9};
    solver_ctx_set_bias(ctx, 5.0);
    solver_ctx_set_branching_weights(ctx, w, 4);
    double ref[4][2][2];
    for (int i = 0; i < 4; i++)
        for (int s = 0; s < 2; s++)
            for (int t = 0; t < 2; t++)
                ref[i][s][t] = BranchingFunction(i, s, t, 0, &ctx->branching_stats);
    solver_ctx_set_angle_precision(ctx, 0.0, 1);   /* dither armed but eps off */
    for (int i = 0; i < 4; i++)
        for (int s = 0; s < 2; s++)
            for (int t = 0; t < 2; t++)
                assert_true(BranchingFunction(i, s, t, 0, &ctx->branching_stats) == ref[i][s][t]);
}

/* ============================================================
 * (2) synthesis-accuracy contract: |theta_q - theta| <= eps
 * ============================================================ */

static void test_quantize_respects_eps(void **state) {
    (void)state;
    /* includes p -> 1 (theta -> pi), where a coarse grid pushes theta_q PAST pi
     * and the sin^2 branch folds back down -- the case the production path can
     * reach whenever a per-variable theta_i drives `value` toward 0. */
    const double values[] = {6.0 / 7.0, 0.5, 0.999, 1.0 - P_N3000_R2, 0.1,
                             1e-9, 1e-6, 0.001};
    const double epss[]   = {0.5, 0.25, 0.1, 0.03, 0.01, 0.001, 1e-5};
    for (size_t vi = 0; vi < sizeof(values) / sizeof(values[0]); vi++) {
        double theta = theta_of(values[vi]);
        for (size_t ei = 0; ei < sizeof(epss) / sizeof(epss[0]); ei++) {
            for (int dither = 0; dither <= 1; dither++) {
                for (int idx = 0; idx < 32; idx++) {
                    double vq = branch_quantize_angle(values[vi], epss[ei], dither, idx);
                    double theta_q = theta_of(vq);
                    /* theta_of() returns the PRINCIPAL branch in [0, pi]. sin^2
                     * is even and pi-symmetric, so a theta_q outside [0, pi]
                     * comes back folded; the accuracy contract must be checked
                     * against the nearest congruent representative. */
                    double d = fabs(theta_q - theta);
                    double folded = fabs((2.0 * M_PI - theta_q) - theta);
                    if (folded < d) d = folded;
                    /* the clamp caps |dtheta| at the extremes, never widens it */
                    assert_true(d <= epss[ei] + 1e-9);
                }
            }
        }
    }
}

/* Monotone convergence: v_q -> v as eps shrinks (both modes). */
static void test_monotone_convergence(void **state) {
    (void)state;
    const double v = 1.0 - P_N3000_R2;
    double prev_sys = 1.0, prev_dit = 1.0;
    for (double eps = 0.1; eps > 1e-7; eps /= 10.0) {
        double e_sys = fabs(branch_quantize_angle(v, eps, 0, 7) - v);
        double e_dit = fabs(branch_quantize_angle(v, eps, 1, 7) - v);
        assert_true(e_sys <= prev_sys + 1e-15);
        assert_true(e_dit <= prev_dit + 1e-15);
        prev_sys = e_sys; prev_dit = e_dit;
    }
    assert_true(prev_sys < 1e-6);
    assert_true(prev_dit < 1e-6);
}

/* ============================================================
 * (3) BOUNDED DECISIONS (§1.7) at every eps -- the fail-loud case
 * ============================================================ */

/* At n=3000, r=2 the angle is theta = 2*asin(sqrt(2/3000)) ~ 0.0516 rad. A grid
 * with eps >= theta/2 rounds it to ZERO: p_flip = 0, a deterministic greedy
 * assignment. The lever must clamp, never return exactly 0 or 1. */
static void test_bounded_decisions_coarse_grid_n3000(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    solver_ctx_set_bias(ctx, BIAS_N3000_R2);
    const double epss[] = {0.5, 0.25, 0.125, 0.0625, 0.03, 0.015, 0.008,
                           0.004, 0.002, 0.001, 1e-6};
    for (size_t ei = 0; ei < sizeof(epss) / sizeof(epss[0]); ei++) {
        for (int dither = 0; dither <= 1; dither++) {
            solver_ctx_set_angle_precision(ctx, epss[ei], dither);
            for (int idx = 0; idx < 256; idx++) {
                /* The (bit_S, bit_T) = (0,0) cell returns the decision value
                 * itself; that is where the (BRANCH_EPS, 1-BRANCH_EPS) bracket
                 * lives. The flipped cells return 1 - value, whose bracket is
                 * only exact up to one ulp (1 - (1 - 1e-9) == 9.9999997e-10),
                 * so those are held to the invariant that actually matters:
                 * STRICTLY inside (0, 1) -- never a forced assignment. */
                double v = BranchingFunction(idx, 0, 0, 0, &ctx->branching_stats);
                assert_true(v >= BRANCH_EPS);
                assert_true(v <= 1.0 - BRANCH_EPS);
                for (int s = 0; s < 2; s++) {
                    for (int t = 0; t < 2; t++) {
                        double p = BranchingFunction(idx, s, t, 0, &ctx->branching_stats);
                        assert_true(p > 0.0);
                        assert_true(p < 1.0);
                        assert_false(isnan(p));
                    }
                }
            }
        }
    }
}

/* The snap-to-zero case is REACHED (the test above would pass vacuously if the
 * grid never actually rounded theta to 0): a coarse systematic grid must pin
 * the flip probability at the clamp floor, not somewhere in the interior. */
static void test_systematic_grid_snaps_to_clamp(void **state) {
    (void)state;
    const double v = 1.0 - P_N3000_R2;             /* theta ~ 0.0516 */
    double vq = branch_quantize_angle(v, 0.5, 0, 11);
    assert_true(vq == 1.0 - BRANCH_EPS);           /* clamped, NOT exactly 1.0 */
    assert_true(1.0 - vq > 0.0);
}

/* Bounded decisions must also hold through the per-variable weights path. */
static void test_bounded_decisions_with_weights(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    double w[8] = {-40.0, -5.0, -0.5, 0.0, 0.5, 5.0, 40.0, 1e3};
    solver_ctx_set_bias(ctx, BIAS_N3000_R2);
    solver_ctx_set_branching_weights(ctx, w, 8);
    for (double eps = 0.5; eps > 1e-7; eps /= 4.0) {
        for (int dither = 0; dither <= 1; dither++) {
            solver_ctx_set_angle_precision(ctx, eps, dither);
            for (int i = 0; i < 8; i++) {
                double p = BranchingFunction(i, 0, 0, 0, &ctx->branching_stats);
                assert_true(p >= BRANCH_EPS && p <= 1.0 - BRANCH_EPS);
                assert_false(isnan(p));
            }
        }
    }
}

/* ============================================================
 * (4) the dither is a deterministic pure function of the index
 * ============================================================ */

static void test_dither_is_deterministic(void **state) {
    (void)state;
    for (int idx = 0; idx < 64; idx++) {
        double a = branch_dither_u(idx);
        double b = branch_dither_u(idx);
        assert_true(a == b);                       /* bit-for-bit reproducible */
        assert_true(a >= -1.0 && a < 1.0);
    }
}

/* Zero-mean-ish and actually varying: the sqrt(n) averaging the lever models
 * only holds if u is not a constant. */
static void test_dither_is_zero_mean_and_varying(void **state) {
    (void)state;
    const int N = 4096;
    double sum = 0.0;
    int distinct = 0;
    double first = branch_dither_u(0);
    for (int i = 0; i < N; i++) {
        double u = branch_dither_u(i);
        sum += u;
        if (u != first) distinct++;
    }
    assert_true(distinct > N / 2);
    assert_true(fabs(sum / N) < 0.05);             /* |mean| well inside 1/sqrt(N) x 3 */
}

/* With dither ON, different variables get different quantized angles; with it
 * OFF they all get the SAME one (the coherent/systematic case). */
static void test_dither_varies_per_index_systematic_does_not(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    solver_ctx_set_bias(ctx, BIAS_N3000_R2);

    solver_ctx_set_angle_precision(ctx, 0.01, 0);
    double v0 = BranchingFunction(0, 0, 0, 0, &ctx->branching_stats);
    for (int i = 1; i < 32; i++)
        assert_true(BranchingFunction(i, 0, 0, 0, &ctx->branching_stats) == v0);

    solver_ctx_set_angle_precision(ctx, 0.01, 1);
    int differ = 0;
    double d0 = BranchingFunction(0, 0, 0, 0, &ctx->branching_stats);
    for (int i = 1; i < 32; i++)
        if (BranchingFunction(i, 0, 0, 0, &ctx->branching_stats) != d0) differ++;
    assert_true(differ >= 30);
}

/* The physics the sweep tests: systematic rounding is COHERENT (the realized
 * radius n*p_q is off by ~n*dp), dithered rounding AVERAGES (~sqrt(n)*dp). */
static void test_dither_averages_radius_error(void **state) {
    (void)state;
    const int n = 3000;
    const double v = 1.0 - P_N3000_R2, eps = 0.01;
    double sys = 0.0, dit = 0.0;
    for (int i = 0; i < n; i++) {
        sys += 1.0 - branch_quantize_angle(v, eps, 0, i);
        dit += 1.0 - branch_quantize_angle(v, eps, 1, i);
    }
    double r_exact = n * P_N3000_R2;               /* == 2 */
    double err_sys = fabs(sys - r_exact) / r_exact;
    double err_dit = fabs(dit - r_exact) / r_exact;
    /* The systematic grid is coherent, so its relative radius error is the full
     * 2*dtheta/theta; the dithered one must be strictly smaller here. */
    assert_true(err_dit < err_sys);
}

/* ============================================================
 * setters: per-phase, and the all-phases convenience
 * ============================================================ */

static void test_per_phase_setters(void **state) {
    solver_ctx_t *ctx = (solver_ctx_t *)*state;
    solver_ctx_set_sat_angle_precision(ctx, 0.1, 0);
    solver_ctx_set_opt_sat_angle_precision(ctx, 0.2, 1);
    solver_ctx_set_opt_angle_precision(ctx, 0.3, 0);
    assert_true(ctx->branching_stats_sat.angle_precision_eps == 0.1);
    assert_int_equal(ctx->branching_stats_sat.angle_precision_dither, 0);
    assert_true(ctx->branching_stats_opt_sat.angle_precision_eps == 0.2);
    assert_int_equal(ctx->branching_stats_opt_sat.angle_precision_dither, 1);
    assert_true(ctx->branching_stats_opt.angle_precision_eps == 0.3);
    assert_int_equal(ctx->branching_stats_opt.angle_precision_dither, 0);

    solver_ctx_set_angle_precision(ctx, 0.05, 1);
    assert_true(ctx->branching_stats_sat.angle_precision_eps == 0.05);
    assert_true(ctx->branching_stats_opt_sat.angle_precision_eps == 0.05);
    assert_true(ctx->branching_stats_opt.angle_precision_eps == 0.05);
    assert_int_equal(ctx->branching_stats_opt.angle_precision_dither, 1);
}

static void test_setters_null_ctx_is_noop(void **state) {
    (void)state;
    solver_ctx_set_angle_precision(NULL, 0.1, 1);
    solver_ctx_set_sat_angle_precision(NULL, 0.1, 1);
    solver_ctx_set_opt_sat_angle_precision(NULL, 0.1, 1);
    solver_ctx_set_opt_angle_precision(NULL, 0.1, 1);
}

int main(void) {
    const struct CMUnitTest tests[] = {
        cmocka_unit_test_setup_teardown(test_defaults_are_off, ap_setup, ap_teardown),
        cmocka_unit_test_setup_teardown(test_golden_6_over_7_preserved, ap_setup, ap_teardown),
        cmocka_unit_test_setup_teardown(test_nonpositive_eps_is_off, ap_setup, ap_teardown),
        cmocka_unit_test_setup_teardown(test_off_matches_baseline_everywhere, ap_setup, ap_teardown),
        cmocka_unit_test(test_quantize_respects_eps),
        cmocka_unit_test(test_monotone_convergence),
        cmocka_unit_test_setup_teardown(test_bounded_decisions_coarse_grid_n3000, ap_setup, ap_teardown),
        cmocka_unit_test(test_systematic_grid_snaps_to_clamp),
        cmocka_unit_test_setup_teardown(test_bounded_decisions_with_weights, ap_setup, ap_teardown),
        cmocka_unit_test(test_dither_is_deterministic),
        cmocka_unit_test(test_dither_is_zero_mean_and_varying),
        cmocka_unit_test_setup_teardown(test_dither_varies_per_index_systematic_does_not, ap_setup, ap_teardown),
        cmocka_unit_test(test_dither_averages_radius_error),
        cmocka_unit_test_setup_teardown(test_per_phase_setters, ap_setup, ap_teardown),
        cmocka_unit_test(test_setters_null_ctx_is_noop),
    };
    return cmocka_run_group_tests(tests, NULL, NULL);
}
